Refactor Specs

waldyr opened the pull request:
WIP for #978 

Beginning refactor our specs so one Smell Detector does not break most tests.

waldyr commented:
@troessner @mvz @chastell 

Is this how the samples folder should look like?

Can I start fixing those tests?

mvz commented:
Since cleaning up everything will be a long process, I would highly recommend doing this one sample file at a time. Baby steps.

troessner commented:
>> Since cleaning up everything will be a long process, I would highly recommend doing this one sample file at a time. Baby steps.

Couldn't agree more. Also "1 file removal = 1 commit" might make the review easier for us :)

waldyr commented:
Sorry for my hastiness. It's been quite some time since my last contribution and I'm in an urge for developing more detectors.

@mvz @troessner 

Deleted `two_smelly_files` and fixed their specs by pointing them to new `smelly`.

troessner commented:
>> Deleted two_smelly_files and fixed their specs by pointing them to new smelly.

Yes, that looks good to me! I think that's exactly the way we need to go about this :)

waldyr commented:
@mvz @troessner 

I've deleted the clean_due_to_masking/dirty_*.rb files.
I also added a configuration/configuration.reek responsible for suppressing the one and only smell smelly.rb has.

Are we also going to create a single clean file? How is it going to be?

troessner commented:
>> Are we also going to create a single clean file? How is it going to be?

How about

```Ruby
# Well documented class.
class Awesome
  def hello(argument)
    say argument
  end
end
```

?

waldyr commented:
@troessner @mvz 

So we have three major files:
- Clean.rb
- Dirty.rb
- Configuration.reek

Buuuuut in reek's tests there are some tests that partially mask a smelly file. Should Dirty.rb reek of two smells and we have two configuration files one for both smells and other to mask one?

troessner commented:
>> We can either add one more smell in Dirty.rb or create a empty configuration file. Wdyt?

Hmmm, good point. Let's go with one additional smell, that's not too much overhead and it is still self-explanatory ;)

waldyr commented:
@troessner @mvz 

Everything is all right?

troessner commented:
>> Everything is alright?

I did a preliminary review and this looks great so far!
How about making the filenames constants as well to prevent fat-fingering and remove duplication?
E.g. "clean.rb" appears multiple times across multiple files.

Other than that I think you should just go on the way you did so far ;)

waldyr commented:
Good idea. I'm also thinking in refactor the examiner spec because the error output wasn't helping me at all with those shared examples. Wdyt?

mvz commented:
Yeah, looks good so far.

mvz commented:
Maybe you should do the examiner refactoring in a next pull request. I'd like to get this one merged.

troessner commented:
Yep, agree with @mvz on the Examiner

waldyr commented:
Should I put the constants in the spec file? I think if I do that I would be hiding where I'm defining them.

Could you guys tell me what more can I do here? I need guidance :P

troessner commented:
Yeah, the spec_helper seems to be the right place.
You can then also:

* look into all the spec/samples_with_* directories and remove / streamline the files in there with what you have done so far. E.g. we can get rid of [this one](https://github.com/troessner/reek/blob/master/spec/samples/source_with_non_ruby_files/uncommunicative_parameter_name.rb) and so on. 
* Kill the whole samples/no_config_file directory
* Streamline "dirty.rb" in samples/masked_by_dotfile

waldyr commented on spec/reek/configuration/configuration_file_finder_spec.rb:
```diff
@@ -25,47 +25,61 @@
+        FileUtils.mkdir(subsubdir)
+        FileUtils.touch(config)
+
+        found = described_class.find(current: subsubdir, home: subsubdir)
```
```
dir
 |-- .reek
 |-- subdir
       |-- subsubdir
```

Why this test does not pass? Do we traverse up from :home? @troessner @mvz

troessner replied:
Not from home, but from the current directory, check out: https://github.com/troessner/reek#configuration-loading

waldyr commented on spec/reek/configuration/configuration_file_finder_spec.rb:
```diff
@@ -83,7 +97,7 @@
     describe '.load_from_file' do
       let(:sample_configuration_path) do
-        SAMPLES_PATH.join('configuration/simple_configuration.reek')
+        SAMPLES_PATH.join('configuration/full_mask.reek')
```
CONFIG_PATH

troessner commented:
@waldyr lets stop here and review & merge, wdyt?

waldyr commented:
Okay. I was looking `spec/samples_with_*` but I didn't find any clean way to remove it.

troessner commented:
I'll do a review of this tomorrow.

troessner commented on spec/samples/configuration/.reek:
```diff
@@ -0,0 +1,9 @@
+    enabled: false
+exclude_paths:
+  - "spec/samples/two_smelly_files/"
+  - "spec/samples/source_with_non_ruby_files/"
```
This file is confusing since it seems like a grabbag for all kind of things. Why do we need it? By whom is it used for what?

waldyr replied:
https://github.com/troessner/reek/pull/992/files#diff-0a2d7be0c2679312a6bcf13434c60172R34

troessner replied:
Then rather empty pls

troessner commented on spec/reek/examiner_spec.rb:
```diff
@@ -47,14 +47,15 @@
-    let(:path) { SAMPLES_PATH.join('all_but_one_masked/masked.reek') }
-    let(:smelly_file) { Pathname.glob(SAMPLES_PATH.join('all_but_one_masked/d*.rb')).first }
+    let(:path) { CONFIG_PATH.join('partial_mask.reek') }
+    let(:smelly_file) { SMELLY_FILE }
```
Why not just use SMELLY_FILE directly like you do below in the specs as well?

troessner commented on spec/reek/examiner_spec.rb:
```diff
@@ -47,14 +47,15 @@
 
   context 'with a fragrant File' do
-    let(:clean_file) { Pathname.glob(SAMPLES_PATH.join('three_clean_files/*.rb')).first }
+    let(:clean_file) { CLEAN_FILE }
```
Same question here.

troessner commented:
1 question and 2 small things and then lgtm!

waldyr commented:
@troessner comments addressed.

@mvz any comments?

troessner commented:
Pls squash as well :)

waldyr commented:
Done!

troessner commented:
Great job @waldyr - merged!
